api/parts/data/exports.js

```diff
@@ -132,45 +169,151 @@ exports.output = function(params, data, filename, type) {
+exports.stream = function(params, stream, filename, type, projection) {
     var headers = {};
+    var listAtEnd = true;
     if (type && contents[type]) {
         headers["Content-Type"] = contents[type];
     }
     headers["Content-Disposition"] = "attachment;filename=" + encodeURIComponent(filename) + "." + type;
+    var paramList = [];
+    if (projection) {
+        for (var k in projection) { //keep order as in projection if given
+            paramList.push(k);
+            listAtEnd = false;
+        }
+    }
     if (params.res.writeHead) {
         params.res.writeHead(200, headers);
-        params.res.write("[");
-        var first = false;
-        stream.on('data', function(doc) {
-            if (!first) {
-                first = true;
-                params.res.write(doc);
+        if (type === "csv") {
+            var head = [];
+            if (listAtEnd === false) {
+                for (let p = 0; p < paramList.length; p++) {
+                    head.push(processCSVvalue(paramList[p]));
+                }
+                params.res.write(head.join(',') + '\r\n');
             }
-            else {
-                params.res.write("," + doc);
+
+            stream.on('data', function(doc) {
+                var values = [];
+                var valuesMap = {};
+                getValues(values, valuesMap, paramList, doc, listAtEnd); // if we have list at end - then we don'thave projection
+
+                for (let p = 0; p < values.length; p++) {
+                    values[p] = processCSVvalue(values[p]);
+                }
+                params.res.write(values.join(',') + '\r\n');
+            });
+
+            stream.once('close', function() {
+                if (listAtEnd) {
+                    for (var p = 0; p < paramList.length; p++) {
+                        head.push(processCSVvalue(paramList[p]));
+                    }
+                    params.res.write(head.join(',') + '\r\n');
+                }
+                params.res.end();
+            });
+        }
+        else if (type === 'xlsx' || type === 'xls') {
+            var xc = xlsx();
+            xc.pipe(params.res);
+            var sheet = xc.sheet("Countly export");
+            if (listAtEnd === false) {
+                sheet.write(paramList);
             }
-        });
+            stream.on('data', function(doc) {
+                var values = [];
+                var valuesMap = {};
+                getValues(values, valuesMap, paramList, doc, listAtEnd);
+                sheet.write(values);
+            });
 
-        stream.once('end', function() {
-            params.res.write("]");
-            params.res.end();
-        });
+            stream.once('close', function() {
+                if (listAtEnd) {
+                    sheet.write(paramList);
+                }
+                sheet.end();
+                xc.finalize();
+            });
+        }
+        else {
```

Comment:
Just out of curiosity, is the purpose of this `else {` branch to handle JSON?

Reply:
yes